instrumentation/opentelemetry-instrumentation-pulsar/src/opentelemetry/instrumentation/pulsar/__init__.py

```diff
@@ -0,0 +1,299 @@
+            ):
+                ctx = propagator.extract(message.properties())
+                span = self._tracer.start_span(
+                    f"{topic} process", ctx, SpanKind.CONSUMER
+                )
+                _enrich_span(
+                    span,
+                    topic,
+                    MessagingOperationValues.RECEIVE,
+                    message.partition_key(),
+                )
+                _enrich_span_with_message(span, message)
+                message._set_current_span(span)
+                with trace.use_span(span, True):
+                    return message_listener(consumer, message, *args, **kwargs)
+
+        bound_arguments.arguments[MESSAGE_LISTENER_ARGUMENT] = wrapper
+        # no need of self
+        args = bound_arguments.args[1:]
+        return super().subscribe(*args, **bound_arguments.kwargs)
+
+
+class InstrumentedProducer(TracerMixin, Producer):
+    _send_signature = inspect.signature(Producer.send)
+
+    def send(self, *args, **kwargs):
+        with self._create_span() as span:
+            args, kwargs = self._before_send(span, *args, **kwargs)
+            message: pulsar.MessageId = super().send(*args, **kwargs)
+            return self._after_send(message, span)
+
+    send.__doc__ = Producer.send.__doc__
+
+    def send_async(self, content, callback, *args, **kwargs):
+        with self._create_span() as span:
+            self._before_send(span, content, callback, *args, **kwargs)
+            callback_context = span.get_span_context()
+
+            @wraps(callback)
+            def wrapper(_response, message_id, *args, **kwargs):
+                with self._tracer.start_as_current_span(
+                    "callback", callback_context
+                ) as span:
+                    self._after_send(message_id, span)
+                    return callback(*args, **kwargs)
+
+            super().send_async(content, wrapper, *args, **kwargs)
+
+    send_async.__doc__ = Producer.send_async.__doc__
+
+    @staticmethod
+    def _after_send(message, span):
+        _enrich_span_with_message_id(span, message)
+        return message
+
+    def _before_send(self, span, *args, **kwargs):
+        bound_arguments = self._send_signature.bind(self, *args, **kwargs)
+        bound_arguments.apply_defaults()
+
+        properties = bound_arguments.arguments[PROPERTIES_KEY] or {}
+        propagator.inject(properties)
+
+        bound_arguments.arguments[PROPERTIES_KEY] = properties
+        _enrich_span(
+            span,
+            self.topic(),
+            operation=MessagingOperationValues.RECEIVE,
+            **bound_arguments.arguments,
+        )
+        return bound_arguments.args[1:], bound_arguments.kwargs
+
+    def _create_span(self):
+        return self._tracer.start_as_current_span(
+            name=_get_span_name("send", self.topic()),
+            kind=trace.SpanKind.PRODUCER,
+        )
+
+
+class InstrumentedMessage(Message, SpanMixin):
```

Comment:
Several classes and constants in this module are public (I mean, their names are not preceded by an underscore). Is this strictly necessary? We prefer to make everything private unless its needed otherwise.

Reply:
Hey there, would you say that only `PulsarInstrumentor` should be public? I will make the change for that